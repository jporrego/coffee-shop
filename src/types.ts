export interface Product {
  id: number;
  name: string;
  price: number;
  description: string;
  img: string;
}

export type StringTouple = [string, string];
