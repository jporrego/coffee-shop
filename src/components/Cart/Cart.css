.icon {
    color: var(--blakc);
    font-size: 3rem;
}