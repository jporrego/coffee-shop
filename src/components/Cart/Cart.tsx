import React, { useEffect, useState } from "react";
import { Product } from "../../types";
import "./Cart.css";
import { BsFillHandbagFill } from "react-icons/bs";
import { IoMdArrowRoundBack } from "react-icons/io";

interface CartProps {
  cartProducts: { product: Product; amount: number }[];
}
const Cart: React.FC<CartProps> = ({ cartProducts }) => {
  const [cartVisible, setCartVisible] = useState(true);

  useEffect(() => {
    setScrolling();
  });

  const getProductAmount = () => {
    let productAmount = 0;
    cartProducts.length > 0 &&
      cartProducts.map((product) => (productAmount += product.amount));
    return productAmount;
  };

  const getTotal = () => {
    let total = 0;
    cartProducts.length > 0 &&
      cartProducts.map(
        (product) => (total += product.product.price * product.amount)
      );
    return total;
  };

  const setScrolling = () => {
    if (cartVisible) {
      document.body.style.overflow = "visible";
    } else {
      document.body.style.overflow = "hidden";
    }
  };

  const cartIcon = () => {
    return (
      <div className="cart-icon" onClick={() => setCartVisible(false)}>
        <BsFillHandbagFill className="icon"></BsFillHandbagFill>
        <div className="cart-icon__product-ammount">{getProductAmount()}</div>
      </div>
    );
  };

  const cartModal = () => {
    return (
      <div className="cart-modal">
        <div className="cart-modal-nav">
          <IoMdArrowRoundBack
            className="btn-back"
            onClick={() => setCartVisible(true)}
          ></IoMdArrowRoundBack>
          <div className="cart-modal-title">My Cart</div>
        </div>
        <div className="cart-modal-product-list">
          {cartProducts.map((product) => (
            <div className="cart-modal-product">
              <div className="cart-modal-product-img">
                <img
                  src={require("../../assets/img/product/" +
                    product.product.img)}
                  alt=""
                />
              </div>
              <div className="cart-modal-product-name">
                {product.product.name}
              </div>
              <div className="cart-modal-product-price">
                ${product.product.price}.00
              </div>
              {product.amount}
              <div className="cart-modal-product-buttons">- 1 + Delete</div>
            </div>
          ))}
        </div>
        <div className="cart-modal-info">
          <div className="cart-modal-amount">{getProductAmount()} items</div>
          <div className="cart-modal-total">${getTotal()}.00</div>
        </div>
      </div>
    );
  };

  return <div className="cart">{cartVisible ? cartIcon() : cartModal()}</div>;
};

export default Cart;
