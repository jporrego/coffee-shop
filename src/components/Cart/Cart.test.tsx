/* eslint-disable testing-library/no-node-access */
import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Cart from "./Cart";
import { Product as ProductInterface } from "../../types";

// set Cart props
type CartProps = {
  cartPoducts: { product: ProductInterface; amount: number }[];
  setCartProducts: typeof jest.fn;
};

const products = [
  {
    product: {
      id: 0,
      brand: "Ovalware",
      name: "Product 1",
      price: 10,
      offer: 0,
      img: "breville.png",
      description: "Product description",
      category: "cold brew",
    },
    amount: 1,
  },
  {
    product: {
      id: 1,
      brand: "Kups",
      name: "Kups",
      price: 149,
      offer: 0,
      description:
        "Does not trap essential oils of your coffee in a paper filter, and makes a beautiful slow-brew. Includes a permanent stainless seel mesh filter.",
      img: "nesspresso-kups.png",
      category: "coffee machines",
    },
    amount: 2,
  },
];

const setCartProducts = jest.fn();

describe("cart", () => {
  it("renders the cart icon", () => {
    const { container } = render(
      <Cart cartProducts={products} setCartProducts={setCartProducts}></Cart>
    );
    expect(container).toMatchSnapshot();
  });

  it("renders the correct amount of products", () => {
    const { container } = render(
      <Cart cartProducts={products} setCartProducts={setCartProducts}></Cart>
    );

    expect(container.textContent).toBe("3");
  });

  it("opens the cart modal", () => {
    const { container } = render(
      <Cart cartProducts={products} setCartProducts={setCartProducts}></Cart>
    );
    userEvent.click(screen.getByText("3"));
    expect(container).toMatchSnapshot();
  });
});
