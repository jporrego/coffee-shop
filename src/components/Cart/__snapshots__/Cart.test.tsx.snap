// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cart opens the cart modal 1`] = `
<div>
  <div
    class="cart"
  >
    <div
      class="cart-modal"
    >
      <div
        class="cart-modal-content"
      >
        <div
          class="cart-modal-nav"
        >
          <svg
            class="btn-back"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M401.4 224h-214l83-79.4c11.9-12.5 11.9-32.7 0-45.2s-31.2-12.5-43.2 0L89 233.4c-6 5.8-9 13.7-9 22.4v.4c0 8.7 3 16.6 9 22.4l138.1 134c12 12.5 31.3 12.5 43.2 0 11.9-12.5 11.9-32.7 0-45.2l-83-79.4h214c16.9 0 30.6-14.3 30.6-32 .1-18-13.6-32-30.5-32z"
            />
          </svg>
          <div
            class="cart-modal-title"
          >
            My Cart
          </div>
        </div>
        <div
          class="cart-modal-product-list"
        >
          <div
            class="cart-product"
          >
            <div
              class="cart-product-img"
            >
              <img
                alt=""
                src="breville.png"
              />
            </div>
            <div
              class="cart-product-name"
            >
              Product 1
            </div>
            <div
              class="cart-product-price"
            >
              $
              10
              .00
            </div>
            <div
              class="cart-product-buttons"
            >
              <div
                class="product-amount-btn"
              >
                -
              </div>
              <div
                class="product-amount"
              >
                1
              </div>
              <div
                class="product-amount-btn"
              >
                +
              </div>
              <svg
                class="delete-icon"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 0h24v24H0V0z"
                  fill="none"
                />
                <path
                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM8 9h8v10H8V9zm7.5-5l-1-1h-5l-1 1H5v2h14V4z"
                />
              </svg>
            </div>
          </div>
          <div
            class="cart-product"
          >
            <div
              class="cart-product-img"
            >
              <img
                alt=""
                src="nesspresso-kups.png"
              />
            </div>
            <div
              class="cart-product-name"
            >
              Kups
            </div>
            <div
              class="cart-product-price"
            >
              $
              149
              .00
            </div>
            <div
              class="cart-product-buttons"
            >
              <div
                class="product-amount-btn"
              >
                -
              </div>
              <div
                class="product-amount"
              >
                2
              </div>
              <div
                class="product-amount-btn"
              >
                +
              </div>
              <svg
                class="delete-icon"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 0h24v24H0V0z"
                  fill="none"
                />
                <path
                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM8 9h8v10H8V9zm7.5-5l-1-1h-5l-1 1H5v2h14V4z"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          class="cart-modal-info"
        >
          <div
            class="cart-modal-amount"
          >
            3
             items
          </div>
          <div
            class="cart-modal-total"
          >
            $
            308
            .00
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`cart renders the cart icon 1`] = `
<div>
  <div
    class="cart"
  >
    <div
      class="cart-icon"
    >
      <svg
        class="icon"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8 1a2 2 0 0 0-2 2v2H5V3a3 3 0 1 1 6 0v2h-1V3a2 2 0 0 0-2-2zM5 5H3.36a1.5 1.5 0 0 0-1.483 1.277L.85 13.13A2.5 2.5 0 0 0 3.322 16h9.355a2.5 2.5 0 0 0 2.473-2.87l-1.028-6.853A1.5 1.5 0 0 0 12.64 5H11v1.5a.5.5 0 0 1-1 0V5H6v1.5a.5.5 0 0 1-1 0V5z"
        />
      </svg>
      <div
        class="cart-icon__product-ammount"
      >
        <div>
          3
        </div>
      </div>
    </div>
  </div>
</div>
`;
